### PR TITLE
Implement Array.prototype.flat/flatMap

### PR DESCRIPTION
**Pull request by davimiku**
This Pull Request fixes part of #36 .

It changes the following:

- Implements Array.prototype.flat
- Implements Array.prototype.flatMap
- Implements abstract function FlattenIntoArray
- Adds test for both builtins

A few things I wasn't sure about:

- I was pretty literal with following the spec, I can tone that back a bit
- The FlattenIntoArray abstract method is supposed to return a "non-negative integer" (it's called recursively and this return value determines when the recursion ends). I had it return a `Value::Integer` for convenience but I didn't know if it should maybe return a primitive `u32` instead
- I have a value compare against `Number::MAX_SAFE_INTEGER` per the spec, but it's a `u32` so couldn't be larger than `2^53` anyways. I didn't test any arrays with huge numbers of elements so there's probably missed edge cases
- I don't think it deals with empty elements in arrays (i.e.`[1, , 3]`), I wasn't sure if those have a representation in Boa yet

### CODE REVIEW

**Review by RageKnify — CHANGES_REQUESTED** [review]
Code looks good. I have one correction and one suggestion that might not be possible.

There's also something else:
```js
let a = [0, 1, 2]
delete a[1]
a.flatMap(x => { return [a] })
```
I think it would return `[0, undefined, 2]` when it should just be `[0, 2]`. (edit: Just noticed that you mentioned knowing that, I'd say to leave coment with `TODO` mentioning that topic so it can be spotted in the future)

As an heads-up you'll need to `merge` `master` into your branch to respect the new lints from Rust 1.51.

File: boa/src/builtins/array/mod.rs [hunk 1]
```diff
@@ -91,6 +92,8 @@ impl BuiltIn for Array {
         .method(Self::every, "every", 1)
         .method(Self::find, "find", 1)
         .method(Self::find_index, "findIndex", 1)
+        .method(Self::flat, "flat", 1)
+        .method(Self::flat_map, "flatMap", 2)
```
[the remark this answers is not in the document]
Comment (davimiku) [reply boa/src/builtins/array/mod.rs]: Fixed!

File: boa/src/builtins/array/mod.rs [hunk 2]
```diff
@@ -885,6 +888,223 @@ impl Array {
...
+        // 4.a. Set depthNum to ToIntegerOrInfinity(depth)
+        // 4.b. If depthNum < 0, set depthNum to 0
+        let depth_num = match depth
+            .unwrap_or(&default_depth)
+            .to_integer_or_infinity(context)?
+        {
+            IntegerOrInfinity::PositiveInfinity => IntegerOrInfinity::PositiveInfinity,
+            IntegerOrInfinity::NegativeInfinity => IntegerOrInfinity::NegativeInfinity,
+            IntegerOrInfinity::Integer(i) if i < 0 => IntegerOrInfinity::Integer(0),
+            IntegerOrInfinity::Integer(i) => IntegerOrInfinity::Integer(i),
```
[the remark this answers is not in the document]
Comment (davimiku) [reply boa/src/builtins/array/mod.rs]: This does work -- thanks! Updated.

**Comment by davimiku** [comment]
I added a check and test for an array with "holes" or empty elements, so that part should be working now. 

I rebased this branch onto master which I think should cover the clippy checks -- but I'm not totally sure (new to Rust and still getting used to the ecosystem).

**Review by RageKnify — APPROVED** [review]

**Review by jasonwilliams — APPROVED** [review]
Looks good I like it. @davimiku im curious to know what your experience was like implementing this. Was it easy/difficult? Did you look at how other methods worked to gain an understanding? Is there something Boa should be doing better to make it easier?

> I was pretty literal with following the spec, I can tone that back a bit

This is a good example of following the spec so we'll keep it around as a use-case. But yeah you don't need to be as literal as long as you can pass test262 there's usually optimizations/shortcuts you can make

> I don't think it deals with empty elements in arrays (i.e.[1, , 3]), I wasn't sure if those have a representation in Boa yet

I think you're right to mention this, it looks like as we parse the array literal to generate the AST we count a hole as undefined and replace it with `undefined`. This happens at [parsing time](https://github.com/boa-dev/boa/blob/master/boa/src/syntax/parser/expression/primary/array_initializer/mod.rs#L68-L71). It looks like smarter engines know the difference between a hole and undefined, this may be a separate issue to investigate and fix.  

> I have a value compare against Number::MAX_SAFE_INTEGER per the spec, but it's a u32 so couldn't be larger than 2^53 anyways. I didn't test any arrays with huge numbers of elements so there's probably missed edge cases 

We can pick this up in Test 262

**Comment by davimiku** [comment]
@jasonwilliams Good question - it was straightforward to find where in the codebase to find these functions and the already implemented functions & tests were helpful as a reference. I think the biggest barrier in my case was outside of Boa, it was getting used to how to read the ECMA Spec and learning Rust syntax.

One thing Boa might be able to do for new contributors is describe how it represents Javascript values -- ex. GcObject vs. Object vs. Array (same as Object?), how prototype methods are called, accessing properties, etc. The other thing that was a bit unclear was the boundary between modules, for example I wasn't sure if I should have used the `array_iterator` module as the iterator for the array builtin functions or just use a 'for' loop (did the latter). I suspect both of those would also clear up more over time.

For the arrays with holes, yeah it sounds like that's probably best as a separate issue. Peeking into Node (V8) internals, there's a different backing store whether the array has holes or not, guessing that there's an iterator used in the array methods that the packed version has an optimization that it doesn't need to check whether the element exists or not.
```
$ node --allow-natives-syntax
> const arr1 = [1, 2, 3]
> %DebugPrint(arr1)
- elements: <FixedArray[3]> [PACKED_SMI_ELEMENTS (COW)]

> const arr2 = [1, , 3]
> %DebugPrint(arr2)
 - elements: <FixedArray[3]> [HOLEY_SMI_ELEMENTS (COW)]
```
